### ISSUE TITLE
Segmentation Fault while training my own dataset

### ISSUE THREAD

**Issue by melvinperera** [google/sling#149]
Hi,

I've been experimenting with SLING to train my own dataset, and this error keeps coming up when running the training script ```train.sh```. This seems to be a similar error to     #115 , but in my case I managed to train the conll dataset successfully. 

I'm using:
- ubuntu 16.04.3
- tensorflow 1.4.0
- python 2.7.12
- gcc 5.4.0 

This is the error log from stdout:
```bash
INFO:tensorflow:Determining the training schedule...
INFO:tensorflow:Training schedule defined!
INFO:tensorflow:Starting training...
2018-03-01 07:16:16.136817: I third_party/syntaxnet/dragnn/core/ops/dragnn_op_kernels.cc:77] Creating new ComputeSessionPool in container handle: shared
2018-03-01 07:16:16.137167: I sling/nlp/parser/trainer/sempar-component.cc:58] lr_lstm: loaded 0 words
2018-03-01 07:16:16.137282: I sling/nlp/parser/trainer/sempar-component.cc:59] lr_lstm: loaded 0 prefixes
2018-03-01 07:16:16.137369: I sling/nlp/parser/trainer/sempar-component.cc:60] lr_lstm: loaded 0 suffixes
2018-03-01 07:16:16.137565: I sling/nlp/parser/trainer/sempar-component.cc:58] rl_lstm: loaded 0 words
2018-03-01 07:16:16.137658: I sling/nlp/parser/trainer/sempar-component.cc:59] rl_lstm: loaded 0 prefixes
2018-03-01 07:16:16.137741: I sling/nlp/parser/trainer/sempar-component.cc:60] rl_lstm: loaded 0 suffixes
2018-03-01 07:16:16.137913: I sling/nlp/parser/trainer/sempar-component.cc:58] ff: loaded 0 words
2018-03-01 07:16:16.138005: I sling/nlp/parser/trainer/sempar-component.cc:59] ff: loaded 0 prefixes
2018-03-01 07:16:16.138087: I sling/nlp/parser/trainer/sempar-component.cc:60] ff: loaded 0 suffixes
./sling/nlp/parser/tools/train.sh: line 243: 16056 Segmentation fault      (core dumped) python sling/nlp/parser/tools/train.py --master_spec="${OUTPUT_FOLDER}/master_spec" --hyperparams="${HYPERPARAMS}" --output_folder=${OUTPUT_FOLDER} --flow=${OUTPUT_FOLDER}/sempar.flow --commons=${COMMONS} --train_corpus=${TRAIN_FILEPATTERN} --dev_corpus=${DEV_FILEPATTERN} --batch_size=${BATCH_SIZE} --report_every=${REPORT_EVERY} --train_steps=${TRAIN_STEPS} --logtostderr
```
I've also attached the following below inside ```files.zip``` in hope that you can replicate and possibly troubleshoot the issue that I have. 

[files.zip](https://github.com/google/sling/files/1770232/files.zip)

- __commons store :__ ```commons```
- __training dataset:__ ```train.rec```
- __test dataset:__ ```test.rec```

Thanks for the help!

**Comment by shishircc** [comment]
It would be great help if you could replicate this issue on your end to help us to get to bottom of why training fails (it’s only one sentence per training data now) 
If same data works on your side that would also give us some clue as present segmentation fault is leaving us no clues.

**Comment by rahul1980** [comment]
I think I know what's happening here, but please allow me to walk through the process.

I peeked at the first few documents in train.rec (via sling.Document). They don't seem to have any tokens (i.e. len(doc.tokens) == 0). This is also reflected in your log (e.g. lr_lstm: loaded 0 words). A zero-sized feature vector is probably what's causing the crash.

Next I printed the documents to confirm the lack of tokens, but surprisingly they do have tokens. So I looked at your commons, and it lacks the symbols required to access the important parts of the document (e.g. /s/document/tokens, /s/document/mention etc). The full list of these symbols is here:

```
 "/s/document",
  "/s/document/mention",
  "/s/document/text",
  "/s/document/theme",
  "/s/document/tokens",
  "/s/phrase",
  "/s/phrase/begin",
  "/s/phrase/evokes",
  "/s/phrase/length",
  "/s/thing",
  "/s/token",
  "/s/token/break",
  "/s/token/index",
  "/s/token/length",
  "/s/token/start",
  "/s/token/text"

```

You can add these symbols to your existing commons via the code snippet below (or just add the symbols above to your existing commons-construction procedure):

```
import sling
commons = sling.Store()
commons.load("/path/to/commons")
required = [
  "/s/document",
  "/s/document/mention",
  "/s/document/text",
  "/s/document/theme",
  "/s/document/tokens",
  "/s/phrase",
  "/s/phrase/begin",
  "/s/phrase/evokes",
  "/s/phrase/length",
  "/s/thing",
  "/s/token",
  "/s/token/break",
  "/s/token/index",
  "/s/token/length",
  "/s/token/start",
  "/s/token/text"
]
for r in required:
  if not r in commons:
    commons.frame({"id" : r})
commons.save("/path/to/new/commons")

```

Can you try rerunning your training procedure with this newly built commons?

**Comment by melvinperera** [comment]
Thanks for the quick reply.

You are right. My commons did not have the symbols required to access the documents.
 
I've updated and rebuilt the commons, but the same Segmentation Fault error pops up. lr_lstm and rl_lstm loaded 0 words as well. 

Using ```commons.save("/path/to/new/commons")``` to update the new commons throws a bazel build error, so I've used the ```build-commons.sh``` file in ```sling/sling/nlp/parser/tools/``` to build the new commons file with the templated schemas found in ```sling/data/nlp/schemas/```. 

I've uploaded the updated commons file below - maybe I've made a mistake in my commons build. Please let me know! Thank you. 

updated commons file: 
[commons.zip](https://github.com/google/sling/files/1774287/commons.zip)


P.S.
```sling/data/nlp/schemas/document-schema.sling``` doesn't have ```token.text``` as a parameter. Was this intentional?

**Comment by rahul1980** [comment]
I also noticed that your train.rec has text valued records instead of binary encoded records, i.e. your record values are textualized documents instead of encoded ones. When you create your train corpus, please use the [Encode method](https://github.com/google/sling/blob/b7fef21fb44962cc008ea9346b3f7d7c5109830c/sling/frame/serialization.h#L380) instead of ToText() to generate the encoded string form. ToText() is just for pretty printing.

Thanks for pointing out the omission in document-schema.sling. I will fix it shortly.

**Comment by rahul1980** [comment]
I just realized that my previous comment may not be fully clear. In your train.rec, the documents are encoded textually, so the reading method (which expects binary encoding) will fail and produce empty documents. This explains the zero vocabulary. So the documents will indeed have to be encoded in binary.

**Comment by shishircc** [comment]
Thanks. Can that be done using python API ?

**Comment by rahul1980** [comment]
Yes. The following worked for me and gave me a word vocab of 335 words.
I have also added a small fix, so please pull from the master and rebuild sling/pyapi:pysling.so before you run the following:

```
import sling
commons = sling.Store()
commons.load("/path/to/new/commons")  # Make sure we use new commons
commons.freeze()
schema = sling.DocumentSchema(commons)

reader = sling.RecordReader("train.rec")
writer = sling.RecordReader("train.new.rec")  # This will replace train.rec in train.sh
for key, val in reader:
  store = sling.Store(commons)
  doc = sling.Document(store.parse(val), schema=schema)
  writer.write(key, doc.frame.data(binary=True, shallow=True))

writer.close()

```

Please also do it for dev.rec and test.rec.

**Comment by melvinperera** [comment]
Yes, that worked indeed. It was because train.rec was encoded as text and not in a binary format. That fixed the problem perfectly. Thank you!

However, there is a new issue now. When training a new corpus, I realised that if my training corpus is reaches past a certain point, a new Segmentation Fault pops up. 

This is the relevant log:
```bash 
INFO:tensorflow:Starting training...
2018-03-05 10:57:08.942218: I third_party/syntaxnet/dragnn/core/ops/dragnn_op_kernels.cc:77] Creating new ComputeSessionPool in container handle: shared
2018-03-05 10:57:08.943015: I sling/nlp/parser/trainer/sempar-component.cc:58] lr_lstm: loaded 108 words
2018-03-05 10:57:08.943157: I sling/nlp/parser/trainer/sempar-component.cc:59] lr_lstm: loaded 0 prefixes
2018-03-05 10:57:08.943270: I sling/nlp/parser/trainer/sempar-component.cc:60] lr_lstm: loaded 213 suffixes
2018-03-05 10:57:08.943378: I sling/nlp/parser/trainer/sempar-component.cc:62] Lexicon OOV: 0
2018-03-05 10:57:08.943484: I sling/nlp/parser/trainer/sempar-component.cc:63] Lexicon normalize digits: 1
2018-03-05 10:57:08.944105: I sling/nlp/parser/trainer/sempar-component.cc:58] rl_lstm: loaded 108 words
2018-03-05 10:57:08.944135: I sling/nlp/parser/trainer/sempar-component.cc:59] rl_lstm: loaded 0 prefixes
2018-03-05 10:57:08.944158: I sling/nlp/parser/trainer/sempar-component.cc:60] rl_lstm: loaded 213 suffixes
2018-03-05 10:57:08.944174: I sling/nlp/parser/trainer/sempar-component.cc:62] Lexicon OOV: 0
2018-03-05 10:57:08.944198: I sling/nlp/parser/trainer/sempar-component.cc:63] Lexicon normalize digits: 1
2018-03-05 10:57:08.944623: I sling/nlp/parser/trainer/sempar-component.cc:58] ff: loaded 0 words
2018-03-05 10:57:08.944660: I sling/nlp/parser/trainer/sempar-component.cc:59] ff: loaded 0 prefixes
2018-03-05 10:57:08.944679: I sling/nlp/parser/trainer/sempar-component.cc:60] ff: loaded 0 suffixes
./sling/nlp/parser/tools/train.sh: line 242:  3849 Segmentation fault      (core dumped) python sling/nlp/parser/tools/train.py --master_spec="${OUTPUT_FOLDER}/master_spec" --hyperparams="${HYPERPARAMS}" --output_folder=${OUTPUT_FOLDER} --flow=${OUTPUT_FOLDER}/sempar.flow --commons=${COMMONS} --train_corpus=${TRAIN_FILEPATTERN} --dev_corpus=${DEV_FILEPATTERN} --batch_size=${BATCH_SIZE} --report_every=${REPORT_EVERY} --train_steps=${TRAIN_STEPS} --logtostderr
```

Attached below are the commons, train.rec and test.rec files.
[files.zip](https://github.com/google/sling/files/1780542/files.zip)

I have a feeling that there's a problem in my corpora somewhere where SLING isn't able to process some documents, but I've been digging in and out of it and can't seem to find what's wrong. Thanks again for your help (it's been excellent so far) and I appreciate any help that you could provide here.

**Comment by rahul1980** [comment]
There are two issues here:
a) One of your documents (document 38, if first document is document 1) has an illegal mention span. Here's the doc:
Document 38
{=#1
  :/s/document
  /s/document/text: "CAC Mid&Small\n(CACMS) and CAC All-Tradable (CACT).\n"
  /s/document/tokens: [{=#2
    :/s/token
    /s/token/index: 0
    /s/token/text: "CAC"
    /s/token/start: 0
    /s/token/length: 3
    /s/token/break: 0
  },
  ...
  }, {=#15
    :/s/token
    /s/token/index: 13
    /s/token/text: "."
    /s/token/start: 49
    /s/token/length: 1
    /s/token/break: 0
  }]
  /s/document/mention: {=#16
    :/s/phrase
    /s/phrase/begin: 10
    /s/phrase/length: 6
    /s/phrase/evokes: {=#17
      :/tr/entity/asset
    }
  }
}

There are only 14 tokens, but the mention is over the range [10, 16). This causes the crash you see.

b) You have no inter-frame links, so our "roles" features have a vocab of zero. I have fixed this in pull request #150  (https://github.com/google/sling/pull/150/files).  You can either pull this commit or locally make the 6-line change in that PR or just wait for the PR to be submitted today and pull from the master.

I have confirmed that if I only use the first 37 documents for training, and include the fix from (b), training proceeds as normal (I ran it for 200 steps).

**Comment by melvinperera** [comment]
That fix absolutely solves the problem and I've managed to get the training working. Thank you so much for your help.
